packages/components/src/components/PasswordCreationField/PasswordCreationField.tsx

```diff
@@ -0,0 +1,315 @@
+import { type PolicyValidationResult } from "@mittwald/password-tools-js/policy";
+import { type RuleValidationResult } from "@mittwald/password-tools-js/rules";
+import { useDebounceValue } from "usehooks-ts";
+import { usePromiseQueue } from "@/components/PasswordCreationField/lib/promiseQueue";
+import { useGeneratePassword } from "@/components/PasswordCreationField/worker/useGeneratePassword";
+import TogglePasswordVisibilityButton from "@/components/PasswordCreationField/components/TogglePasswordVisibilityButton/TogglePasswordVisibilityButton";
+
+const validationDebounceMilliseconds = 200;
+
+export const defaultPasswordCreationPolicy = Policy.fromDeclaration({
+  minComplexity: 3,
+  rules: [
+    {
+      ruleType: RuleType.length,
+      min: 12,
+    },
+    {
+      ruleType: RuleType.hibp,
+    },
+    {
+      identifier: "special",
+      ruleType: RuleType.charPool,
+      charPools: ["special"],
+    },
+    {
+      identifier: "numbers",
+      ruleType: RuleType.charPool,
+      charPools: ["numbers"],
+    },
+  ],
+});
+
+export interface PasswordCreationFieldProps
+  extends PropsWithChildren<
+      Omit<Aria.TextFieldProps, "children" | "value" | "defaultValue"> &
+        Partial<Pick<Aria.FieldErrorRenderProps, "validationErrors">>
+    >,
+    FlowComponentProps<HTMLInputElement> {
+  value?: string;
+  defaultValue?: string;
+  validationPolicy?: Policy;
+}
+
+export interface ResolvedPolicyValidationResult extends PolicyValidationResult {
+  isEmptyValueValidation: boolean;
+  isValid: boolean;
+  ruleResults: RuleValidationResult[];
+}
+
+/**
+ * @flr-generate all
+ * @flr-clear-props-context
+ */
+export const PasswordCreationField = flowComponent(
+  "PasswordCreationField",
+  (props) => {
+    const {
+      children,
+      value: valueFromProps,
+      defaultValue,
+      className,
+      ref,
+      isDisabled,
+      onChange,
+      isInvalid: invalidFromProps,
+      validationPolicy = defaultPasswordCreationPolicy,
+      isRequired,
+      ...rest
+    } = props;
+
+    const promiseQueue = usePromiseQueue();
+    const translate = useLocalizedStringFormatter(locales);
+    const generatePassword = useGeneratePassword(validationPolicy);
+
+    const [isPasswordRevealed, setIsPasswordRevealed] = useState(false);
+    const [isLoading, setIsLoading] = useState(false);
+
+    const initialPolicyValidationState: ResolvedPolicyValidationResult = {
+      isEmptyValueValidation: true,
+      isValid: false,
+      complexity: {
+        min: validationPolicy.minComplexity,
+        actual: 4 as ComplexityScore,
+        warning: null,
+      },
+      ruleResults: [],
+    };
+
+    const [policyValidationResult, setPolicyValidationResult] = useState(
+      initialPolicyValidationState,
+    );
+
+    const stateFromValidationResult = getStateFromLatestPolicyValidationResult(
+      policyValidationResult,
+    );
```

Comment:
Instead of getting the state on every render, you could do this only when a new `policyValidationResult` is calculated, and put it directly into the state.